---
title: "Analysis of SW with 3D heatmap"
author: "Анатолий Сорокин"
date: '`r format(Sys.time(), "%d.%m.%Y")`'
output:
  pdf_document:
    keep_tex: yes
    number_sections: yes
  html_document:
    keep_md: yes
params:
  version: !r system("git describe --long --dirty --abbrev=10  --tags  --always", intern=TRUE)
header-includes:
- \usepackage[T2A]{fontenc}
- \usepackage[utf8]{inputenc}
- \usepackage[english,russian]{babel}
- \usepackage{grffile}
- \usepackage{rotating}
- \usepackage{caption}
- \usepackage{longtable}
- \usepackage{lscape}
---
```{r loadPackages, include=FALSE, cache=FALSE}
## load additional packages in this chunk
library(pander)
library(knitr)
library(ggplot2)
library(mmnet)
library(RCurl)
library(metagenomeSeq)
library(gplots)
library(xtable)
library(data.table)
library(RJSONIO)
library(plyr)
library(pathview)
library(stringr)
```
```{r setup, include=FALSE, cache=FALSE}
## This chunk should contain global configuration commands.
## Use this to set knitr options and related things. Everything
## in this chunk will be included in an appendix to document the
## configuration used.
#output <- opts_knit$get("rmarkdown.pandoc.to")

## By default R code is only included in HTML versions of the report
## (where it can be collapsed). You can generate a PDF version
## using rmarkdown::pdf_document to get a copy for print. Extensive
## chunks of R code may or may not be desired in /hat setting. If you
## want them simply change the following arguments to `echo = TRUE`.
## In either case the default can be overwritten for individual chunks.
#opts_chunk$set(echo = output=="html")
#opts_chunk$set(warning = output=="html")
#opts_chunk$set(message = output=="html")

## Cache options
opts_chunk$set(cache=TRUE)

## Figure options
## Set default figure format
#options(reportmd.figure.format=params$format)

## Set 'hide.fig.code' to FALSE to include code chunks that
## produce Figures in the output. Note that this affects all chunks
## that provide a figure caption.
opts_chunk$set(hold=TRUE, hide.fig.code=FALSE)

## Set up default plotting options for different formats.
## These can be overwritten for individual chunks
#interactiveFig()
#screenFig()
#printFig()

## Pander options
panderOptions("digits", 3)
panderOptions("table.split.table", 160)

## Configure Figure and Table lables
#options(figcap.prefix = "Figure", figcap.sep = ":", figcap.prefix.highlight = "**")
#options(tabcap.prefix = "Table", tabcap.sep = ":", tabcap.prefix.highlight = "**")

## Install required knitr hooks
#installHooks()
old.o <- options(warn = -1)
```

```{r functions, include=TRUE,child='mmnet.functions.Rmd'}
```

```
{r load.annotation.1,cache=TRUE}
load(file = 'mgrast.pathview.Rdata')
```

```
{r fetch.annotation,cache=TRUE}
ids<-mdt$MG.RAST.ID
annot<-lapply(ids,myGetMgrastAnnotation,webkey=key)
ind<-which(!sapply(annot,function(.x)grepl("Download\\s+complete", .x[nrow(.x), 1])))
ind
```

# Analysis of matches
```
{r analyse.matches}
mlist<-dir(path = './tmp',pattern = 'mgm4714659.*')

mannot<-lapply(mlist,function(.x){fread(paste0('ghead -n -1 tmp/',.x),sep='\t',header = TRUE)})
matches<-matrix(0,nrow = length(mannot),ncol = length(mannot))
for(i in 1:length(mannot)) matches[i,i]<-dim(mannot[[i]])[1]
for(i in 2:length(mannot)-1){
for(j in (i+1):length(mannot)){
matches[i,j]<-length(which(!is.na(match(mannot[[i]]$`hit m5nr id (md5sum)`,mannot[[j]]$`hit m5nr id (md5sum)`))))
matches[j,i]<-length(which(!is.na(match(mannot[[j]]$`hit m5nr id (md5sum)`,mannot[[i]]$`hit m5nr id (md5sum)`))))
}
}
rownames(matches)<-mlist
colnames(matches)<-mlist

xtable(matches)
xtable(cor(matches))
```

# process data
```
{r load.fdata.from.file,cache=FALSE}
#flist<-dir(path = './tmp',pattern = '*.ko$')
flist<-dir(path = './tmp',pattern = '*.3.fseed$')
cat(paste(flist,collapse = '\n'))
#flist<-flist[-length(flist)]
fannot<-lapply(flist,function(.x){fread(paste0('ghead -n -1 ./tmp/',.x),sep='\t',header = TRUE)})

```

```
{r load.kodata.from.file,cache=FALSE}
klist<-dir(path = './tmp',pattern = '^m.*.ko$')
cat(paste(klist,collapse = '\n'))
#flist<-flist[-length(flist)]
ko<-lapply(klist,function(.x){fread(paste0('ghead -n -1 ./tmp/',.x),sep='\t',header = TRUE)})

```


```
{r load.sdata.from.file,cache=FALSE}
slist<-dir(path = './tmp',pattern = '*.3.seed$')
cat(paste(slist,collapse = '\n'))
#slist<-slist[-length(slist)]
if(length(slist)!=length(flist)) stop('Length of functional and specie annotation should match\n')
sannot<-lapply(slist,function(.x){fread(paste0('ghead -n -1 ./tmp/',.x),sep='\t',header = TRUE)})

```

```
{r merge}
nms<-gsub('.fseed$','',flist)
res<-list()
kres<-list()
for(i in 1:length(fannot)){
  f<-fannot[[i]]
  #f$`query sequence id`<-gsub('\\|KO$','',f$`query sequence id`)
  s<-sannot[[i]]
  #s$`query sequence id`<-gsub('\\|SEED$','',s$`query sequence id`)
  d.k<-unique(ko[[i]][,list(`hit m5nr id (md5sum)`,`semicolon separated list of annotations`)])
  d.k1<-d.k[,list(`semicolon separated list of annotations`,ko=unlist(gsub('accession=\\[K([0-9]+)\\].*','K\\1',unlist(str_split(`semicolon separated list of annotations`,';'))))),by=.(`hit m5nr id (md5sum)`)]
names(d.k1)<-c('md5','annotation','ko')

  kres[[nms[i]]]<-list(ab=d.k1,name=nms[i])
  d.merge<-mergeAnnots(f,s)
  d.uspfun<-expandNamesDT(d.merge)
  d.ab<-getAbundanceMD5FromDT(d.uspfun)
  res[[nms[i]]]<-list(ab=d.ab,name=nms[i])

  #d.m.t<-table(d.merge$fun,d.merge$sp)
  cat(paste(i,nms[i],'\n'))
}
```

```
{r make.d.res}
d.res<-ldply(.data = res,
             .fun = function(.x){
               ab<-.x$ab;
               ab$mgid=.x$name;
               return(ab)
               })
names(d.res)
d.kres<-unique(ldply(.data = kres,
.fun = function(.x){
ab<-.x$ab;
return(ab)
}))
names(d.kres)

```

```
{r aggregate}
dcast(setDT(d.res),usp+ufun+md5 ~ mgid,value.var = 'sum',fill = 0)->d.bm
names(d.bm)
d.sp<-aggregate(.~usp,as.data.frame(d.bm)[,-c(2,3)],FUN = sum)
d.fun<-aggregate(.~ufun,as.data.frame(d.bm)[,-c(1,3)],FUN = sum)

```


```{r make.pwlist1}
pwlist<-sort(unique(c('00020','00062','00561','00564','00620','00640','00650','00660','00680','00720','00790','00920','02024','02025','05111','00130','00190','00400','00860','00910','01053','01057','02010','02020')))
```

```
{r save.data}
save(pwlist,d.bm,mdt,d.res,d.kres,fannot,sannot,ko,file = 'pathview.Rdata')
```

```{r load.rdata,cache=FALSE}
load(file = 'd.bm.Rdata')
#load(file = 'pathview.Rdata')
```

# Nitrate removal
## Nitrosomonas
```{r Nitrosomonas.fun,fig.width=8.5,fig.height=10.5,results='asis'}
sp.l<-'Nitrosomonas'
d<-getSpecieFromAbundMD5(d.bm,sp = sp.l,aggregate = FALSE)
d5<-d[,list(m5=unlist(str_split(md5,',')),mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3),by=.(usp,ufun,md5)]
dk5<-unique(merge(d5,d.kres,all=FALSE,by.x='m5',by.y='md5')[,.(m5,usp,ufun,mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3,ko)])
adk5<-aggregate(.~ko,as.data.frame(dk5[,-c(1:3)]),FUN=sum)
rownames(adk5)<-adk5$ko
gdk5<-adk5[,2:6]
names(gdk5)<-c('MFC1_p','MFC3_p','MFC1_a','MFC3_a','inflowSW')
gdk5[,5]<-adk5[,1+which(mdt$MFCID==names(gdk5)[5])]
for(i in 1:4) gdk5[,i]<-rowSums(adk5[,1+which(mdt$MFCID==names(gdk5)[i])])


for(pid in pwlist){
pv.out <- pathview(gene.data = adk5[,c(1,8,3,10)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(1,8,3,10)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(1,8,3,10)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(1,8,3,10)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = adk5[,c(4,11,9,6)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(4,11,9,6)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(4,11,9,6)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(4,11,9,6)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = gdk5, pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(gdk5)))))
  pv.out <- pathview(gene.data = log2(gdk5+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(gdk5+1))))))
}
obj<-as.matrix(d[,-(1:3)])
rownames(obj)<-d$ufun
colnames(obj)<-mdt$MGN
res<-plotHeatmap(obj,100,trace = "none", col = heatmapCols,main='Nitrosomonas functions, \nnorm and log-transformed')
showTable(res,'Nitrosomonas')
plotHeatmap(obj,100,norm = FALSE,trace = "none", col = heatmapCols,main='Nitrosomonas functions \nlog-transformed')
plotSP(d.bm[,-3],sp = 'Nitrosomonas')
```

## Nitrosococcus
```{r Nitrosococcus.fun,fig.width=8.5,fig.height=10.5,results='asis'}
sp.l<-'Nitrosococcus'
d<-getSpecieFromAbundMD5(d.bm,sp = sp.l,aggregate = FALSE)
d5<-d[,list(m5=unlist(str_split(md5,',')),mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3),by=.(usp,ufun,md5)]
dk5<-unique(merge(d5,d.kres,all=FALSE,by.x='m5',by.y='md5')[,.(m5,usp,ufun,mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3,ko)])
adk5<-aggregate(.~ko,as.data.frame(dk5[,-c(1:3)]),FUN=sum)
rownames(adk5)<-adk5$ko
gdk5<-adk5[,2:6]
names(gdk5)<-c('MFC1_p','MFC3_p','MFC1_a','MFC3_a','inflowSW')
gdk5[,5]<-adk5[,1+which(mdt$MFCID==names(gdk5)[5])]
for(i in 1:4) gdk5[,i]<-rowSums(adk5[,1+which(mdt$MFCID==names(gdk5)[i])])


for(pid in pwlist){
pv.out <- pathview(gene.data = adk5[,c(1,8,3,10)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(1,8,3,10)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(1,8,3,10)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(1,8,3,10)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = adk5[,c(4,11,9,6)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(4,11,9,6)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(4,11,9,6)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(4,11,9,6)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = gdk5, pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(gdk5)))))
  pv.out <- pathview(gene.data = log2(gdk5+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(gdk5+1))))))
}
obj<-as.matrix(d[,-(1:3)])
rownames(obj)<-d$ufun
colnames(obj)<-mdt$MGN
res<-plotHeatmap(obj,100,trace = "none", col = heatmapCols,main='Nitrosococcus functions, \nnorm and log-transformed')
showTable(res,'Nitrosococcus')
plotHeatmap(obj,100,norm = FALSE,trace = "none", col = heatmapCols,main='Nitrosococcus functions \nlog-transformed')
plotSP(d.bm[,-3],sp = 'Nitrosococcus')
```

## Nitrospina
```
{r Nitrospina.fun,fig.width=8.5,fig.height=10.5,results='asis'}
sp.l<-'Nitrospina'
d<-getSpecieFromAbundMD5(d.bm,sp = sp.l,aggregate = FALSE)
d5<-d[,list(m5=unlist(str_split(md5,',')),mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3),by=.(usp,ufun,md5)]
dk5<-unique(merge(d5,d.kres,all=FALSE,by.x='m5',by.y='md5')[,.(m5,usp,ufun,mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3,ko)])
adk5<-aggregate(.~ko,as.data.frame(dk5[,-c(1:3)]),FUN=sum)
rownames(adk5)<-adk5$ko
gdk5<-adk5[,2:6]
names(gdk5)<-c('MFC1_p','MFC3_p','MFC1_a','MFC3_a','inflowSW')
gdk5[,5]<-adk5[,1+which(mdt$MFCID==names(gdk5)[5])]
for(i in 1:4) gdk5[,i]<-rowSums(adk5[,1+which(mdt$MFCID==names(gdk5)[i])])


for(pid in pwlist){
pv.out <- pathview(gene.data = adk5[,c(1,8,3,10)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(1,8,3,10)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(1,8,3,10)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(1,8,3,10)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = adk5[,c(4,11,9,6)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(4,11,9,6)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(4,11,9,6)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(4,11,9,6)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = gdk5, pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(gdk5)))))
  pv.out <- pathview(gene.data = log2(gdk5+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(gdk5+1))))))
}
obj<-as.matrix(d[,-(1:3)])
rownames(obj)<-d$ufun
colnames(obj)<-mdt$MGN
res<-plotHeatmap(obj,100,trace = "none", col = heatmapCols,main='Nitrospina functions, \nnorm and log-transformed')
showTable(res,'Nitrospina')
plotHeatmap(obj,100,norm = FALSE,trace = "none", col = heatmapCols,main='Nitrospina functions \nlog-transformed')
plotSP(d.bm[,-3],sp = 'Nitrospina')
```

## Nitrospira
```
{r Nitrospira.fun,fig.width=8.5,fig.height=10.5,results='asis'}
sp.l<-'Nitrospira'
d<-getSpecieFromAbundMD5(d.bm,sp = sp.l,aggregate = FALSE)
d5<-d[,list(m5=unlist(str_split(md5,',')),mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3),by=.(usp,ufun,md5)]
dk5<-unique(merge(d5,d.kres,all=FALSE,by.x='m5',by.y='md5')[,.(m5,usp,ufun,mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3,ko)])
adk5<-aggregate(.~ko,as.data.frame(dk5[,-c(1:3)]),FUN=sum)
rownames(adk5)<-adk5$ko
gdk5<-adk5[,2:6]
names(gdk5)<-c('MFC1_p','MFC3_p','MFC1_a','MFC3_a','inflowSW')
gdk5[,5]<-adk5[,1+which(mdt$MFCID==names(gdk5)[5])]
for(i in 1:4) gdk5[,i]<-rowSums(adk5[,1+which(mdt$MFCID==names(gdk5)[i])])


for(pid in pwlist){
pv.out <- pathview(gene.data = adk5[,c(1,8,3,10)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(1,8,3,10)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(1,8,3,10)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(1,8,3,10)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = adk5[,c(4,11,9,6)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(4,11,9,6)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(4,11,9,6)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(4,11,9,6)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = gdk5, pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(gdk5)))))
  pv.out <- pathview(gene.data = log2(gdk5+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(gdk5+1))))))
}
obj<-as.matrix(d[,-(1:3)])
rownames(obj)<-d$ufun
colnames(obj)<-mdt$MGN
res<-plotHeatmap(obj,100,trace = "none", col = heatmapCols,main='Nitrospira functions, \nnorm and log-transformed')
showTable(res,'Nitrospira')
plotHeatmap(obj,100,norm = FALSE,trace = "none", col = heatmapCols,main='Nitrospira functions \nlog-transformed')
plotSP(d.bm[,-3],sp = 'Nitrospira')
```

## Paracoccus
```{r Paracoccus.fun,fig.width=8.5,fig.height=10.5,results='asis'}
sp.l<-'Paracoccus'
d<-getSpecieFromAbundMD5(d.bm,sp = sp.l,aggregate = FALSE)
d5<-d[,list(m5=unlist(str_split(md5,',')),mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3),by=.(usp,ufun,md5)]
dk5<-unique(merge(d5,d.kres,all=FALSE,by.x='m5',by.y='md5')[,.(m5,usp,ufun,mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3,ko)])
adk5<-aggregate(.~ko,as.data.frame(dk5[,-c(1:3)]),FUN=sum)
rownames(adk5)<-adk5$ko
gdk5<-adk5[,2:6]
names(gdk5)<-c('MFC1_p','MFC3_p','MFC1_a','MFC3_a','inflowSW')
gdk5[,5]<-adk5[,1+which(mdt$MFCID==names(gdk5)[5])]
for(i in 1:4) gdk5[,i]<-rowSums(adk5[,1+which(mdt$MFCID==names(gdk5)[i])])


for(pid in pwlist){
pv.out <- pathview(gene.data = adk5[,c(1,8,3,10)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(1,8,3,10)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(1,8,3,10)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(1,8,3,10)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = adk5[,c(4,11,9,6)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(4,11,9,6)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(4,11,9,6)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(4,11,9,6)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = gdk5, pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(gdk5)))))
  pv.out <- pathview(gene.data = log2(gdk5+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(gdk5+1))))))
}
obj<-as.matrix(d[,-(1:3)])
rownames(obj)<-d$ufun
colnames(obj)<-mdt$MGN
res<-plotHeatmap(obj,100,trace = "none", col = heatmapCols,main='Paracoccus functions, \nnorm and log-transformed')
showTable(res,'Paracoccus')
plotHeatmap(obj,100,norm = FALSE,trace = "none", col = heatmapCols,main='Paracoccus functions \nlog-transformed')
plotSP(d.bm[,-3],sp = 'Paracoccus')
```

## Thiobacillus
```{r Thiobacillus.fun,fig.width=8.5,fig.height=10.5,results='asis'}
sp.l<-'Thiobacillus'
d<-getSpecieFromAbundMD5(d.bm,sp = sp.l,aggregate = FALSE)
d5<-d[,list(m5=unlist(str_split(md5,',')),mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3),by=.(usp,ufun,md5)]
dk5<-unique(merge(d5,d.kres,all=FALSE,by.x='m5',by.y='md5')[,.(m5,usp,ufun,mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3,ko)])
adk5<-aggregate(.~ko,as.data.frame(dk5[,-c(1:3)]),FUN=sum)
rownames(adk5)<-adk5$ko
gdk5<-adk5[,2:6]
names(gdk5)<-c('MFC1_p','MFC3_p','MFC1_a','MFC3_a','inflowSW')
gdk5[,5]<-adk5[,1+which(mdt$MFCID==names(gdk5)[5])]
for(i in 1:4) gdk5[,i]<-rowSums(adk5[,1+which(mdt$MFCID==names(gdk5)[i])])


for(pid in pwlist){
pv.out <- pathview(gene.data = adk5[,c(1,8,3,10)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(1,8,3,10)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(1,8,3,10)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(1,8,3,10)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = adk5[,c(4,11,9,6)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(4,11,9,6)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(4,11,9,6)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(4,11,9,6)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = gdk5, pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(gdk5)))))
  pv.out <- pathview(gene.data = log2(gdk5+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(gdk5+1))))))
}
obj<-as.matrix(d[,-(1:3)])
rownames(obj)<-d$ufun
colnames(obj)<-mdt$MGN
res<-plotHeatmap(obj,100,trace = "none", col = heatmapCols,main='Thiobacillus functions, \nnorm and log-transformed')
showTable(res,'Thiobacillus')
plotHeatmap(obj,100,norm = FALSE,trace = "none", col = heatmapCols,main='Thiobacillus functions \nlog-transformed')
plotSP(d.bm[,-3],sp = 'Thiobacillus')
```


## Nitrosospira
```{r Nitrosospira.fun,fig.width=8.5,fig.height=10.5,results='asis'}
sp.l<-'Nitrosospira'
d<-getSpecieFromAbundMD5(d.bm,sp = sp.l,aggregate = FALSE)
d5<-d[,list(m5=unlist(str_split(md5,',')),mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3),by=.(usp,ufun,md5)]
dk5<-unique(merge(d5,d.kres,all=FALSE,by.x='m5',by.y='md5')[,.(m5,usp,ufun,mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3,ko)])
adk5<-aggregate(.~ko,as.data.frame(dk5[,-c(1:3)]),FUN=sum)
rownames(adk5)<-adk5$ko
gdk5<-adk5[,2:6]
names(gdk5)<-c('MFC1_p','MFC3_p','MFC1_a','MFC3_a','inflowSW')
gdk5[,5]<-adk5[,1+which(mdt$MFCID==names(gdk5)[5])]
for(i in 1:4) gdk5[,i]<-rowSums(adk5[,1+which(mdt$MFCID==names(gdk5)[i])])


for(pid in pwlist){
pv.out <- pathview(gene.data = adk5[,c(1,8,3,10)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(1,8,3,10)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(1,8,3,10)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(1,8,3,10)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = adk5[,c(4,11,9,6)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(4,11,9,6)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(4,11,9,6)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(4,11,9,6)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = gdk5, pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(gdk5)))))
  pv.out <- pathview(gene.data = log2(gdk5+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(gdk5+1))))))
}
obj<-as.matrix(d[,-(1:3)])
rownames(obj)<-d$ufun
colnames(obj)<-mdt$MGN
res<-plotHeatmap(obj,100,trace = "none", col = heatmapCols,main='Nitrosospira functions, \nnorm and log-transformed')
showTable(res,'Nitrosospira')
plotHeatmap(obj,100,norm = FALSE,trace = "none", col = heatmapCols,main='Nitrosospira functions \nlog-transformed')
plotSP(d.bm[,-3],sp = 'Nitrosospira')
```

## Nitrosolobus
```
{r Nitrosolobus.fun,fig.width=8.5,fig.height=10.5,results='asis'}
sp.l<-'Nitrosolobus'
d<-getSpecieFromAbundMD5(d.bm,sp = sp.l,aggregate = FALSE)
d5<-d[,list(m5=unlist(str_split(md5,',')),mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3),by=.(usp,ufun,md5)]
dk5<-unique(merge(d5,d.kres,all=FALSE,by.x='m5',by.y='md5')[,.(m5,usp,ufun,mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3,ko)])
adk5<-aggregate(.~ko,as.data.frame(dk5[,-c(1:3)]),FUN=sum)
rownames(adk5)<-adk5$ko
gdk5<-adk5[,2:6]
names(gdk5)<-c('MFC1_p','MFC3_p','MFC1_a','MFC3_a','inflowSW')
gdk5[,5]<-adk5[,1+which(mdt$MFCID==names(gdk5)[5])]
for(i in 1:4) gdk5[,i]<-rowSums(adk5[,1+which(mdt$MFCID==names(gdk5)[i])])


for(pid in pwlist){
pv.out <- pathview(gene.data = adk5[,c(1,8,3,10)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(1,8,3,10)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(1,8,3,10)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(1,8,3,10)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = adk5[,c(4,11,9,6)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(4,11,9,6)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(4,11,9,6)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(4,11,9,6)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = gdk5, pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(gdk5)))))
  pv.out <- pathview(gene.data = log2(gdk5+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(gdk5+1))))))
}
obj<-as.matrix(d[,-(1:3)])
rownames(obj)<-d$ufun
colnames(obj)<-mdt$MGN
res<-plotHeatmap(obj,100,trace = "none", col = heatmapCols,main='Nitrosolobus functions, \nnorm and log-transformed')
showTable(res,'Nitrosolobus')
plotHeatmap(obj,100,norm = FALSE,trace = "none", col = heatmapCols,main='Nitrosolobus functions \nlog-transformed')
plotSP(d.bm[,-3],sp = 'Nitrosolobus')
```

## Nitrosopumilus
```{r Nitrosopumilus.fun,fig.width=8.5,fig.height=10.5,results='asis'}
sp.l<-'Nitrosopumilus'
d<-getSpecieFromAbundMD5(d.bm,sp = sp.l,aggregate = FALSE)
d5<-d[,list(m5=unlist(str_split(md5,',')),mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3),by=.(usp,ufun,md5)]
dk5<-unique(merge(d5,d.kres,all=FALSE,by.x='m5',by.y='md5')[,.(m5,usp,ufun,mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3,ko)])
adk5<-aggregate(.~ko,as.data.frame(dk5[,-c(1:3)]),FUN=sum)
rownames(adk5)<-adk5$ko
gdk5<-adk5[,2:6]
names(gdk5)<-c('MFC1_p','MFC3_p','MFC1_a','MFC3_a','inflowSW')
gdk5[,5]<-adk5[,1+which(mdt$MFCID==names(gdk5)[5])]
for(i in 1:4) gdk5[,i]<-rowSums(adk5[,1+which(mdt$MFCID==names(gdk5)[i])])


for(pid in pwlist){
pv.out <- pathview(gene.data = adk5[,c(1,8,3,10)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(1,8,3,10)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(1,8,3,10)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(1,8,3,10)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = adk5[,c(4,11,9,6)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(4,11,9,6)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(4,11,9,6)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(4,11,9,6)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = gdk5, pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(gdk5)))))
  pv.out <- pathview(gene.data = log2(gdk5+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(gdk5+1))))))
}
obj<-as.matrix(d[,-(1:3)])
rownames(obj)<-d$ufun
colnames(obj)<-mdt$MGN
res<-plotHeatmap(obj,100,trace = "none", col = heatmapCols,main='Nitrosopumilus functions, \nnorm and log-transformed')
showTable(res,'Nitrosopumilus')
plotHeatmap(obj,100,norm = FALSE,trace = "none", col = heatmapCols,main='Nitrosopumilus functions \nlog-transformed')
plotSP(d.bm[,-3],sp = 'Nitrosopumilus')
```

## Nitrososphaera
```
{r Nitrososphaera.fun,fig.width=8.5,fig.height=10.5,results='asis'}
sp.l<-'Nitrososphaera'
d<-getSpecieFromAbundMD5(d.bm,sp = sp.l,aggregate = FALSE)
d5<-d[,list(m5=unlist(str_split(md5,',')),mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3),by=.(usp,ufun,md5)]
dk5<-unique(merge(d5,d.kres,all=FALSE,by.x='m5',by.y='md5')[,.(m5,usp,ufun,mgm4714659.3,mgm4714661.3,mgm4714663.3,mgm4714665.3,mgm4714667.3,mgm4714669.3,mgm4714671.3,mgm4714673.3,mgm4714675.3,mgm4714677.3,mgm4714679.3,ko)])
adk5<-aggregate(.~ko,as.data.frame(dk5[,-c(1:3)]),FUN=sum)
rownames(adk5)<-adk5$ko
gdk5<-adk5[,2:6]
names(gdk5)<-c('MFC1_p','MFC3_p','MFC1_a','MFC3_a','inflowSW')
gdk5[,5]<-adk5[,1+which(mdt$MFCID==names(gdk5)[5])]
for(i in 1:4) gdk5[,i]<-rowSums(adk5[,1+which(mdt$MFCID==names(gdk5)[i])])


for(pid in pwlist){
pv.out <- pathview(gene.data = adk5[,c(1,8,3,10)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(1,8,3,10)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(1,8,3,10)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.plankton"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(1,8,3,10)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = adk5[,c(4,11,9,6)+1], pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(adk5[,c(4,11,9,6)+1]))),cpd=1))
  pv.out <- pathview(gene.data = log2(adk5[,c(4,11,9,6)+1]+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko.anode"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(adk5[,c(4,11,9,6)+1]+1)))),cpd=1))
  pv.out <- pathview(gene.data = gdk5, pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(gdk5)))))
  pv.out <- pathview(gene.data = log2(gdk5+1), pathway.id = pid,
                     species = "ko", out.suffix = paste0(sp.l,".log2.ko"), kegg.native = T,
                     limit = list(gene=range(as.vector(as.matrix(log2(gdk5+1))))))
}
obj<-as.matrix(d[,-(1:3)])
rownames(obj)<-d$ufun
colnames(obj)<-mdt$MGN
res<-plotHeatmap(obj,100,trace = "none", col = heatmapCols,main='Nitrososphaera functions, \nnorm and log-transformed')
showTable(res,'Nitrososphaera')
plotHeatmap(obj,100,norm = FALSE,trace = "none", col = heatmapCols,main='Nitrososphaera functions \nlog-transformed')
plotSP(d.bm[,-3],sp = 'Nitrososphaera')
```




# Приложения {.tabset}
## Функции
```{r functions, eval=FALSE, include=TRUE}
```


## Конфигурация R
```{r setup, eval=FALSE}
```

## Версии
### Версия документа
```{r docVersion, echo=FALSE, results='asis', cache=FALSE}
cat(params$version)
```

### Session Info
```{r sessionInfo, echo=FALSE, results='asis', class='text', warning=FALSE}
pander(devtools::session_info())
options(old.o)
```

