<h1 id="advanced-metagenomic-sequence-analysis-in-r">Advanced metagenomic Sequence Analysis in R</h1>
<h2 id="manual-for-version-1-revision-1">Manual for version 1, revision 1</h2>
<h3 id="august-2017">August, 2017</h3>
<p><em>Askarbek Orakov<sup>1,2,*</sup>, Nazgul Sakenova<sup>1,2</sup>, Anatoly Sorokin<sup>3,4</sup> and Igor Goryanin<sup>1,5,6</sup></em></p>
<p><em><sup>1</sup>Okinawa Institute of Science and Technology, Onna-son, Japan, <sup>2</sup>School of Science and Technology, Nazarbayev University, Astana, Kazakhstan, <sup>3</sup>Institute of Cell Biophysics RAS, Pushchino, Russia, <sup>4</sup>Laboratory of Ion and Molecular Physics, Moscow Institute of Physics and Technology, Moscow, Russia. <sup>5</sup>University of Edinburgh, School of Informatics, Edinburgh, United Kingdom, <sup>6</sup>Tianjin Institute of Industrial Biotechnology, Biodesign centre, Tianjin, China.</em></p>
<p>Code can be found on GitHub: <a href="https://github.com/Askarbek-orakov/ASAR" class="uri">https://github.com/Askarbek-orakov/ASAR</a></p>
<h2 id="outline">Outline:</h2>
<ol style="list-style-type: decimal">
<li><p>Summary</p></li>
<li><p>Introduction</p></li>
<li><p>Results</p></li>
<li><p>Parameters</p></li>
<li><p>Installation</p></li>
<li><p>Data Preparation</p></li>
<li><p>Manual download of files</p></li>
<li><p>How to cite us</p></li>
<li><p>References</p></li>
</ol>
<h2 id="summary">Summary</h2>
<p><strong>What is it?</strong> The functional and taxonomic analysis is the critical step in understanding the interspecies interaction within the biological community. At the moment these types of analysis are run separately which makes results difficult for interpretation. Here we present ASAR interactive tool for simultaneous analysis of the metagenomic data along three dimensions: taxonomy, function, metagenome.</p>
<p><strong>What is required from a user?</strong> The user needs to have R and some R packages installed in the user’s computer. Additionally, terminal is needed to run the BASH script that will download, process and save data in the appropriate format to be read by the application. The annotation files from MG-RAST are used as an example data. In order to use data from MG-RAST only project ID and user webkey are needed from the user.</p>
<p><strong>Why should I use it?</strong> Advantages of the tool are 1) Integrated functional and taxonomic analysis; 2) Comparative analysis of KEGG pathway enrichments; 3) KEGG Pathway Maps; 4) User-friendly interface.</p>
<h2 id="introduction">Introduction</h2>
<p>This application implements two general analyses. First, it builds 3D dataset consisting of axes of taxonomy, functions and metagenome samples [1,2]. Second, through KEGG metabolic pathways analysis consists of a comparative analysis of pathways enrichment and visualization of pathways themselves [3].</p>
<p>Since taxonomic and functional annotations have many groups at several levels and metagenome samples are numerous, two main data manipulations are implemented. First, “Selection” involves selecting one or several separate groups in each axis at a certain level in functions and taxonomy. Second, “Aggregation” involves selecting a level lower than “Selection” level at which selected data should be aggregated to groups of this level by summing read counts. “Aggregation” of metagenomes is done by averaging metagenomes with same defined name.</p>
<p>There is the sample data to explore the app, and to use your own data, please, read the section “Data preparation”.</p>
<h2 id="results">Results</h2>
<p><strong><em>3D dataset (Function &amp; Taxonomy &amp; Metagenomes) Interactive Heatmaps</em></strong></p>
<p>Heatmaps contain dendrograms to the left and above from heatmap and column and row names below and to the right from the heatmap. Dendrograms are generated with “hclust” function in R and color key upleft to heatmaps represents the distribution of color in the heatmap. The value of the cell in the heatmap can be viewed by locating mouse cursor above that cell. The value of the cell is the log2 value of read count for that cell.</p>
<div class="figure">
<img src="media/image1.png" width="684" height="396" />

</div>
<div class="figure">
<img src="media/image2.png" width="696" height="396" />

</div>
<div class="figure">
<img src="media/image3.png" width="660" height="408" />

</div>
<p><strong><em>KEGG Pathways Heatmap</em></strong></p>
<p><img src="media/image4.png" width="672" height="372" />In addition to characteristics explained for 3D dataset heatmaps, KEGG Pathways Heatmap has Standard Deviation cutoff explained in parameters section for SD cutoff.</p>
<p><strong><em>KEGG Pathway</em></strong></p>
<p><img src="media/image5.png" width="684" height="300" />KEGG Pathway shows the metabolic pathway image for selected KEGG pathway and color enzymes with known expression in metagenomes. Each rectangle representing enzyme is horizontally divided into a number of selected metagenomes, where coloration of partitions corresponds to values of selected metagenomes(i.e. Partitions are colored from left to right as an order of selected metagenomes). Values of enrichment represent percent contribution of selected taxon from total read count of the whole metagenome to this KEGG Orthology(enzyme). The color key will represent a variation from zero to maximum value among KO’s in the pathway.</p>
<h2 id="parameters">Parameters</h2>
<p><strong><em>Metagenome Selection</em></strong></p>
<p>Function vs Taxonomy (F/T) heatmap requires selection of single metagenome, while all others allow selection of multiple metagenomes(see figure 6). Names of metagenomes shown for selection can be changed by selecting a specific column of metadata, rows of which will be used as metagenome names (see figure 7). In order to consider several metagenomes as one, these metagenomes should be given the same name in selected metadata column. In this case abundances of reads of metagenomes with same names are averaged in new metagenome under shared name before any abundances processing steps.</p>
<div class="figure">
<img src="media/image6.png" width="672" height="324" />

</div>
<div class="figure">
<img src="media/image8.png" width="684" height="503" />

</div>
<p><strong><em>Taxonomy and Function Selection</em></strong></p>
<p><img src="media/image10.png" width="701" height="468" />There are 8 choices for taxonomy level, where &quot;root&quot; means all domains and 7 levels from “domain” to “strain”. At all levels except &quot;root&quot; taxon selection will give a list of all available taxons at that level and multiple taxon selections are possible. At last, second taxon level is used to aggregate selected taxons at that level. Functions are selected with the same principle with the only difference that it has 5 levels instead of 8 (see figure 8).</p>
<p><strong><em>Pathway Selection for Building KEGG Pathway</em></strong></p>
<p>The list of KEGG Pathways available for current selection of metagenomes and taxons is displayed for selection. Selecting Pathway will send the request to build KEGG Pathway and may take up to several seconds depending on the number of genes in the pathway.</p>
<p><strong><em>Heatmap Height Selection</em></strong></p>
<p>Default height of heatmaps is 20 pixels per row and is adjustable through slider input parameter, which displays and sets the height of a single row in pixels.</p>
<p><strong><em>Image Download</em></strong></p>
<p>Every heatmap can be downloaded by typing a user-defined file name in “Enter file name” text input parameter and selecting image format (PNG or PDF) and subsequently pressing the Download button. KEGG map image is downloaded in the same way but without defining file name. Alternatively, map image can be saved by clicking the right mouse button (usual browser functionality), where user can define filename.</p>
<p><strong><em>Standard deviation cutoff for KEGG Orthology terms</em></strong></p>
<p>KEGG Pathways Heatmap and KEGG Pathway have the parameter called “SD cutoff for KO terms” which defines the value of Standard Deviation for individual KO’s among all selected metagenomes and is adjustable. This value cuts off all KO’s with SD less than that value from Heatmap.</p>
<p><strong><em>Parameters in Metadata Tab</em></strong></p>
<p>Selection of column of metadata displays all column of metadata and allows to use rows of these columns as names of metagenomes in metagenome selection parameter. Metadata values are editable and new columns with a name specified by user can be added to the metadata. There are three types of new column user can select, namely “integer”, “double” and “character”. Pushing “Save” button updates default RData file with current dataset. Pushing “Save a new column” button will update metadata but current Rdata should be saved and run in new session. When metagenome names from some column of metadata are used identically named metagenome data will be averaged and analyzed as single metagenome. This is how aggregation of metagenomes is done.</p>
<p><strong><em>Parameters in Settings Tab</em></strong></p>
<p>Upload function can be used to upload Rdata files generated previously and Save function can be used to save current state of loaded dataset. Settings tab has entries for changing default values of functional and taxonomic levels. Palette of colors used for coloring heatmaps can also be selected. Pressing “Save changes” button will save these default parameters for next sessions.</p>
<h2 id="installation">Installation</h2>
<p>R Packages from CRAN:</p>
<ol style="list-style-type: decimal">
<li><p>Package ‘shiny’ version 1.0.3</p></li>
<li><p>Package ‘ggplot2’ version 2.2.1</p></li>
<li><p>Package ‘gplots’ version 3.0.1</p></li>
<li><p>Package ‘data.table’ version 1.10.4</p></li>
<li><p>Package ‘plyr’ version 1.8.4</p></li>
<li><p>Package ‘stringr’ version 1.2.0</p></li>
<li><p>Package ‘shinythemes’ version 1.1.1</p></li>
<li><p>Package ‘matrixStats’ version 0.52.2</p></li>
<li><p>Package ‘png’ version 0.1-7</p></li>
<li><p>Package ‘devtools’ version 1.13.2</p></li>
<li><p>Package ‘rhandsontable’ version 0.3.4.6</p></li>
<li><p>Package ‘RColorBrewer’ version 1.1-2</p></li>
</ol>
<p>Bioconductor:</p>
<ol style="list-style-type: decimal">
<li><p>Package ‘mmnet’ version 1.15.0-1</p></li>
<li><p>Package ‘pathview’ version 1.14.0</p></li>
<li><p>Package ‘biomformat’ version 1.2.0</p></li>
<li><p>Package ‘KEGGREST’ version 1.14.1</p></li>
<li><p>Package ‘limma’ version 3.30.13</p></li>
</ol>
<p>GitHub:</p>
<ol style="list-style-type: decimal">
<li>Package ‘d3heatmap’ by “Alanocallaghan/d3heatmap”</li>
</ol>
<p>To run the app on your local machine:</p>
<ol style="list-style-type: decimal">
<li><p>Download RStudio/R</p></li>
<li><p>Run these commands below in the console:</p></li>
</ol>
<p>install.packages(c(&quot;shiny&quot;,&quot;ggplot2&quot;,&quot;gplots&quot;,&quot;RColorBrewer&quot;,&quot;data.table&quot;,&quot;plyr&quot;,&quot;stringr&quot;,&quot;shinythemes&quot;,&quot;matrixStats&quot;,&quot;png&quot;,&quot;devtools&quot;,&quot;rhandsontable&quot;))</p>
<p>## try http:// if https:// URLs are not supported<br />
source(&quot;https://bioconductor.org/biocLite.R&quot;)<br />
biocLite(&quot;mmnet&quot;)</p>
<p>biocLite(&quot;pathview&quot;)</p>
<p>biocLite(&quot;biomformat&quot;)</p>
<p>biocLite(&quot;KEGGREST&quot;)</p>
<p>biocLite(&quot;limma&quot;)</p>
<p>install_github(&quot;Alanocallaghan/d3heatmap&quot;)</p>
<ol style="list-style-type: decimal">
<li>Run this command in the console: shiny::runGitHub(&quot;ASAR&quot;, &quot;askarbek&quot;)</li>
</ol>
<h2 id="data-preparation">Data Preparation</h2>
<p>Preparation of data from MG-RAST only requires project ID and webkey to be given to BASH script which subsequently downloads and processes all required files and generates Rdata file that is directly used by the application.</p>
<h2 id="manual-download-of-files">Manual Download of Files</h2>
<p>The list of required input files:</p>
<ol style="list-style-type: decimal">
<li><p>Functional annotations file either by KEGG or SEED</p></li>
<li><p>Taxonomic annotations file either by KEGG or SEED</p></li>
<li><p>KEGG Orthology file</p></li>
<li><p>Biom file</p></li>
<li><p>Metadata</p></li>
</ol>
<p>Our app uses MG-RAST annotations as an example. MG-RAST has both public and private projects which can be downloaded as it is described in its <a href="ftp://ftp.metagenomics.anl.gov/data/manual/mg-rast-manual.pdf"><em>manual</em></a>. There are two ways of downloading files from MG-RAST and prepare them for the input.</p>
<p>The first way is to download files directly from MG-RAST website, while second way is to download files through API or other command line tools, such as a terminal. In the case of the former way, you will have to rename files manually, while in the case of the latter way, given code will download files automatically and rename them automatically.</p>
<p>1. Download files directly from MG-RAST website as follows.</p>
<ol style="list-style-type: lower-alpha">
<li>Select a project.</li>
</ol>
<div class="figure">
<img src="media/image16.png" alt="Screen Shot 2017-07-19 at 13.50.01.png" width="684" height="164" />
<p class="caption">Screen Shot 2017-07-19 at 13.50.01.png</p>
</div>
<ol style="list-style-type: lower-alpha">
<li>Select a sample.</li>
</ol>
<div class="figure">
<img src="media/image17.png" alt="Screen Shot 2017-07-19 at 13.52.04.png" width="690" height="148" />
<p class="caption">Screen Shot 2017-07-19 at 13.52.04.png</p>
</div>
<ol style="list-style-type: lower-alpha">
<li>Go to Download.</li>
</ol>
<div class="figure">
<img src="media/image18.png" alt="Screen Shot 2017-07-19 at 13.55.34.png" width="689" height="212" />
<p class="caption">Screen Shot 2017-07-19 at 13.55.34.png</p>
</div>
<ol style="list-style-type: lower-alpha">
<li>Download functional annotation file either by SEED or KEGG by selecting “function” for Annotation Type and either “KEGG” or “SEED” for Data Source and rename them by adding “.fkegg” or “.fseed” respectively. Examples: “mgm4714675.3.fkegg” and “mgm4714675.3.fseed”.</li>
</ol>
<p><img src="media/image19.png" alt="Screen Shot 2017-07-18 at 14.02.21.png" width="564" height="97" /><img src="media/image20.png" alt="Screen Shot 2017-07-18 at 13.43.32.png" width="565" height="101" /></p>
<ol style="list-style-type: lower-alpha">
<li>Download taxonomic annotation file either by SEED or KEGG by selecting “organism” for Annotation Type and either “KEGG” or “SEED” for Data Source and rename them by adding “.kegg” or “.seed” respectively. Examples: “mgm4714675.3.kegg” and “mgm4714675.3.seed”.</li>
</ol>
<div class="figure">
<img src="media/image21.png" alt="Screen Shot 2017-07-18 at 14.04.37.png" width="559" height="97" />
<p class="caption">Screen Shot 2017-07-18 at 14.04.37.png</p>
</div>
<div class="figure">
<img src="media/image22.png" alt="Screen Shot 2017-07-18 at 13.43.58.png" width="554" height="100" />
<p class="caption">Screen Shot 2017-07-18 at 13.43.58.png</p>
</div>
<ol style="list-style-type: lower-alpha">
<li>Download KEGG Orthology file by selecting “ontology” for Annotation Type and “KO” for Data Source and rename the file by adding at the end “.ko”. Example: “mgm4714675.3.ko”.</li>
</ol>
<div class="figure">
<img src="media/image23.png" alt="Screen Shot 2017-07-18 at 13.20.50.png" width="552" height="105" />
<p class="caption">Screen Shot 2017-07-18 at 13.20.50.png</p>
</div>
<ol style="list-style-type: lower-alpha">
<li>Biome file can be downloaded only from MG-RAST API/command line tools (see below)</li>
</ol>
<p>g. Download metadata file by entering a project and pressing file icon as shown below and rename the file as “jobs.tsv”.</p>
<p>1.Press icon shown below.</p>
<div class="figure">
<img src="media/image24.png" alt="Screen Shot 2017-07-19 at 14.00.07.png" width="525" height="113" />
<p class="caption">Screen Shot 2017-07-19 at 14.00.07.png</p>
</div>
<p>2. Select all.</p>
<div class="figure">
<img src="media/image25.png" alt="Screen Shot 2017-07-19 at 14.02.15.png" width="509" height="375" />
<p class="caption">Screen Shot 2017-07-19 at 14.02.15.png</p>
</div>
<p>3. Press the icon shown below to download metadata.</p>
<div class="figure">
<img src="media/image26.png" alt="Screen Shot 2017-07-19 at 14.08.22.png" width="595" height="100" />
<p class="caption">Screen Shot 2017-07-19 at 14.08.22.png</p>
</div>
<p>2. Download files through API or other command line tools, such as a terminal.</p>
<ol style="list-style-type: decimal">
<li><p>You can find how to download files through API by clicking <a href="http://api.metagenomics.anl.gov/api.html"><em>this link.</em></a></p></li>
<li><p>An example of how to download files through terminal is shown below.</p></li>
</ol>
<p>Webkey will be needed to download files from private projects. To get your webkey in MG-RAST, press “ show webkey” as indicated below.</p>
<div class="figure">
<img src="media/image27.png" alt="Screen Shot 2017-07-19 at 11.16.45.png" width="272" height="140" />
<p class="caption">Screen Shot 2017-07-19 at 11.16.45.png</p>
</div>
<p>Open Terminal and run this chunk of code after modifying webkey and metagenome ID (as marked by red squares) to download your files.</p>
<div class="figure">
<img src="media/image28.png" alt="Screen Shot 2017-07-19 at 13.19.26.png" width="642" height="57" />
<p class="caption">Screen Shot 2017-07-19 at 13.19.26.png</p>
</div>
<ol style="list-style-type: decimal">
<li>Functional annotation by SEED</li>
</ol>
<p>curl -H &quot;auth: your_webkey_comes_here&quot; -H 'Accept-Encoding: gzip,deflate' &quot;http://api-pql.metagenomics.anl.gov/1/annotation/similarity/mgm4714675.3?source=SEED&amp;type=function&amp;identity=60&amp;length=15&quot; -o mgm4714679.3.fseed</p>
<ol style="list-style-type: decimal">
<li>Taxonomic annotation by SEED</li>
</ol>
<p>curl -H &quot;auth: your_webkey_comes_here&quot; -H 'Accept-Encoding: gzip,deflate' &quot;http://api.metagenomics.anl.gov/1/annotation/similarity/mgm4714675.3?source=SEED&amp;type=organism&amp;identity=60&amp;length=15&quot; -o mgm4714675.3.seed</p>
<ol style="list-style-type: decimal">
<li>Functional annotation by KEGG</li>
</ol>
<p>curl -H &quot;auth: your_webkey_comes_here&quot; -H 'Accept-Encoding: gzip,deflate' &quot;http://api-pql.metagenomics.anl.gov/1/annotation/similarity/mgm4714675.3?source=KEGG&amp;type=function&amp;identity=60&amp;length=15&quot; -o mgm4714679.3.fkegg</p>
<ol style="list-style-type: decimal">
<li>Taxonomic annotation by KEGG</li>
</ol>
<p>curl -H &quot;auth: your_webkey_comes_here&quot; -H 'Accept-Encoding: gzip,deflate' &quot;http://api-pql.metagenomics.anl.gov/1/annotation/similarity/mgm4714675.3?source=KEGG&amp;type=organism&amp;identity=60&amp;length=15&quot; -o mgm4714679.3.kegg</p>
<ol style="list-style-type: decimal">
<li>KEGG Orthology file</li>
</ol>
<p>curl -H &quot;auth: your_webkey_comes_here&quot; -H 'Accept-Encoding: gzip,deflate' &quot;http://api-pql.metagenomics.anl.gov/1/annotation/similarity/mgm4714675.3?source=KO&amp;type=ontology&amp;identity=60&amp;length=15&quot; -o mgm4714675.3.ko</p>
<ol style="list-style-type: decimal">
<li>Biome file:</li>
</ol>
<p>curl -H &quot;auth: your_webkey_comes_here&quot; -H 'Accept-Encoding: gzip,deflate' &quot;http://api-pql.metagenomics.anl.gov/1/matrix/organism?id=mgm4714675.3&amp;id=mgm4714661.3&amp;id=mgm4714663.3&amp;source=SEED&amp;group_level=strain&amp;result_type=abundance&amp;hit_type=all&amp;identity=60&amp;length=15&quot; -o mgm.biome</p>
<ol style="list-style-type: decimal">
<li>Metadata</li>
</ol>
<p>Metadata is created by user.</p>
<p><strong>Creating .RData</strong></p>
<p>Put all five files into the same directory.<br />
You may use this app by<br />
A. Exploring the pre-loaded example data set. This is a pre-loaded Metagome Samples taken from swine waste example for exploring the app's features.</p>
<p>B. Upload your own data that is either<br />
C. Uploading an .RData file containing your data that was previously downloaded from the app session.</p>
<h2 id="cite-us">Cite us:</h2>
<p>Askarbek N. Orakov, Nazgul K. Sakenova, Anatoly Sorokin and Igor Goryanin. 2017. “ASAR: ”.</p>
<h2 id="references">References:</h2>
<p>[1]Keegan, K. P., Glass, E. M., &amp; Meyer, F. (2016). MG-RAST, a metagenomics service for analysis of microbial community structure and function. Microbial Environmental Genomics (MEG), 207-233.</p>
<p>[2]Overbeek, R., Begley, T., Butler, R. M., Choudhuri, J. V., Chuang, H. Y., Cohoon, M., ... &amp; Fonstein, M. (2005). The subsystems approach to genome annotation and its use in the project to annotate 1000 genomes. <em>Nucleic acids research</em>, <em>33</em>(17), 5691-5702.</p>
<p>[3]Kanehisa, M., Sato, Y., Kawashima, M., Furumichi, M., &amp; Tanabe, M. (2016). KEGG as a reference resource for gene and protein annotation. <em>Nucleic acids research</em>, <em>44</em>(D1), D457-D462.</p>
